lib/chef/knife/core/bootstrap_context.rb

```diff
@@ -172,11 +172,11 @@ def config_content
           end
 
           unless chef_config[:file_cache_path].nil?
-            client_rb << "file_cache_path \"#{chef_config[:file_cache_path]}\"\n"
+            client_rb << "file_cache_path \"#{ChefConfig::PathHelper.escapepath(ChefConfig::Config.var_chef_dir(windows: false))}/cache\"\n"
```

Comment:
we've got the windows_bootstrap_context which handles bootstrapping a windows node.  this should properly be called unix_bootstrap_context or something, but history.

Reply:
whats a bit more concerning is this is now hardcoding the last bit of this path.

Reply:
we probably need to introduce new config variables `unix_bootstrap_file_cache_path` and `unix_bootstrap_file_backup_path` which should default to these values and allow the user to override them.  then should probably do the same for windows.

Reply:
@lamont-granquist I have tried adding the new variable and separate definitions  for defining  `chef_dir`, `root_dir` & `cache_path` for Linux  in [config.rb](https://github.com/chef/chef/blob/master/chef-config/lib/chef-config/config.rb). But I got stuck [here](https://github.com/chef/chef/blob/master/chef-config/lib/chef-config/config.rb#L372), where we need to get a hone path of the remote node. I think we need to update `ChefUtils` to get the object of remove OS.

Please let me know your suggestion on this.

Reply:
@snehaldwivedi ChefUtils does have a little bit of train in it and can get the remote OS, but it is for target mode, knife bootstrap has its own use of train.

`ChefConfig::Mixin::TrainTransport` is used for target mode, not bootstrap.
`ChefUtils::DSL::TrainHelpers` ultimately uses the train connection off the run_context for target mode
`Chef::Knife::Bootstrap::TrainConnector` is used for knife bootstrap.

The amount of complexity that ChefConfig has around switching for windows or linux, local mode or target mode, and changing path names for the official or community distributions is already a lot and easy to break one of the many scenarios. Trying to add knife bootstrap or not knife bootstrap to that list is over the top.

For just knife bootstrap use in chef-config we could add something like this:
```
default(:unix_bootstrap_file_cache_path) do
    PathHelper.join("/var", ChefUtils::Dist::Infra::DIR_SUFFIX, "cache", windows: false)
end

default(:windows_bootstrap_file_cache_path) do
  PathHelper.join("C:", ChefUtils::Dist::Infra::DIR_SUFFIX, "cache", windows: true)
end
```

If someone needed something special (like probably we leave C: hardcoded) then they could put a different `windows_bootstrap_file_cache_path` in their config.rb.

Then we mostly need to go through `bootstrap_context.rb` and `windows_bootstrap_context.rb` and make them use the appropriate new variable instead of `file_cache_path`, and update the tests. And do a little manual testing.